Sources/SwiftSyntax/SyntaxText.swift

```diff
@@ -0,0 +1,226 @@
+
+  public subscript(bounds: Range<Index>) -> SyntaxText {
+    if isEmpty && bounds.isEmpty { return self }
+    return SyntaxText(
+      baseAddress: buffer.baseAddress! + bounds.lowerBound,
+      count: bounds.count)
+  }
+
+  public subscript(position: Index) -> Element {
+    unsafeAddress { buffer.baseAddress!.advanced(by: position) }
+  }
+}
+
+extension SyntaxText: Hashable {
+  public static func ==(lhs: SyntaxText, rhs: SyntaxText) -> Bool {
+    if lhs.buffer.count != rhs.buffer.count {
+      return false
+    }
+    if lhs.buffer.baseAddress == rhs.buffer.baseAddress {
+      return true
+    }
+    return compareMemory(lhs.baseAddress!, rhs.baseAddress!, lhs.count)
+  }
+
+  public func hash(into hasher: inout Hasher) {
+    hasher.combine(bytes: .init(buffer))
+  }
+}
+
+extension SyntaxText: ExpressibleByStringLiteral {
+  public init(stringLiteral value: StaticString) { self.init(value) }
+  public init(unicodeScalarLiteral value: StaticString) { self.init(value) }
+  public init(extendedGraphemeClusterLiteral value: StaticString) { self.init(value) }
+}
+
+extension SyntaxText: CustomStringConvertible {
+  public var description: String { String(syntaxText: self) }
+}
+
+extension SyntaxText: CustomDebugStringConvertible {
+  public var debugDescription: String { description.debugDescription }
+}
+
+extension String {
+  /// Creates a `String` from a `SyntaxText`.
+  ///
+  /// Ill-formed UTF-8 sequences in `syntaxText` are replaced with the Unicode
+  /// replacement character `\u{FFFD}`.
+  @_spi(Testing)
+  public init(syntaxText: SyntaxText) {
+    guard !syntaxText.isEmpty else {
+      self = ""
+      return
+    }
+    let count = syntaxText.count
+    if #available(macOS 11.0, iOS 14.0, watchOS 7.0, tvOS 14.0, *) {
+      self.init(unsafeUninitializedCapacity: count) { strBuffer in
+        copyMemory(strBuffer.baseAddress!, syntaxText.baseAddress!, count)
+        return count
+      }
+    } else {
+      self.init(decoding: syntaxText, as: UTF8.self)
+    }
+  }
+
+  /// Runs `body` with a `SyntaxText` that refers the contiguous memory of this
+  /// string. Like `String.withUTF8(_:)`, this may mutates the string if this
+  /// string was not contiguous.
+  @_spi(Testing)
+  public mutating func withSyntaxText<R>(
+    _ body: (SyntaxText) throws -> R
+  ) rethrows -> R {
+    try withUTF8 { utf8 in
+      try body(SyntaxText(baseAddress: utf8.baseAddress, count: utf8.count))
+    }
+  }
+}
+
+private func compareMemory(
+  _ s1: UnsafePointer<UInt8>, _ s2: UnsafePointer<UInt8>, _ count: Int
+) -> Bool {
+  assert(count > 0)
+#if canImport(Darwin)
+  return Darwin.memcmp(s1, s2, count) == 0
+#elseif canImport(Glibc)
+  return Glibc.memcmp(s1, s2, count) == 0
+#else
+  return UnsafeBufferPointer(start: s1, count: count)
+    .elementsEqual(UnsafeBufferPointer(start: s2, count: count))
+#endif
+}
+
+private func copyMemory(
+  _ dst: UnsafeMutablePointer<UInt8>, _ src: UnsafePointer<UInt8>, _ count: Int
```

Comment:
A little QoL nit: Please add `from`, `to`, and `count` labels here. I always have to put a little extra thought into the order of arguments to memcpy/memcmp/memmove.

Reply:
I removed `copyMemory()` because I think `buffer.initialize(from:)` is enough.